NGINX: Do not redirect /api/v1/ URLs from HTTP to HTTPS.

Joelute opened the pull request:
Following the article [Your API Shouldn't Redirect HTTP to HTTPS](https://jviide.iki.fi/http-redirects), API requests through HTTP are very dangerous and exposes your API key to the network in plaintext. Many of these requests are unintentional as API clients follow the requests silently without raising any error/warning.

We should follow the fail fast principle and outright reject unencrypted API calls to our servers, providing the developers a visible way to quickly patch this security issue.

Currently we are simply redirecting all HTTP requests through NGINX, which is dangerous as we are sending sensitive secrets through plain/text. We want to stop redirecting for API requests and handle them with an middleware in Django with a possibility of implementing an auto key revoker later on. These changes employs the following:

- Allow unencrypted API requests through reverse proxy to Django
- Detect for unencrypted API requests in our backend
- Reject and return a standard error message letting clients know of the situation 

This will allow us and developers using our API to handle the requests in a more secure manner.

CZO: [api-design](https://chat.zulip.org/#narrow/stream/378-api-design/topic/Your.20API.20Shouldn.27t.20Redirect.20HTTP.20to.20HTTPS), [backend](https://chat.zulip.org/#narrow/stream/3-backend/topic/Don't.20redirect.20API.20from.20HTTP.20to.20HTTPS)

Fixes: #30264.

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Joelute committed:
NGINX: Do not redirect /api/v1/ URLs from HTTP to HTTPS.

We currently redirect all unencrypted HTTP requests to HTTPS in NGINX.
While this is reasonable for browser-based environments, which
generally segregate HTTP and HTTPS cookies (ensuring that client
secrets are not leaked in the HTTP request), this poses an security
risk for API clients.  Specifically, an API client which is configured
to auto-follow redirects may send an API key in cleartext, and then
_silently_ follow the 301 redirect provided by NGINX, leaving the user
blind to the fact that their credentials were exposed in cleartext.

Stop redirecting unencrypted API requests, and provide clients with a
clear message alerting them to the fact that credentials were
potentially exposed.

Fixes: #30264.

alexmv commented on zerver/lib/exceptions.py:
```diff
@@ -31,6 +31,7 @@ class ErrorCode(Enum):
     REQUEST_CONFUSING_VAR = auto()
     INVALID_API_KEY = auto()
     INVALID_ZOOM_TOKEN = auto()
+    INVALID_REQUEST = auto()
```
Adding this into the middle of the list _changes all of the later IDs_, making them mean something else to clients that rely on them.

You can only add new values to the end of this list.

Please also add a comment to this to clarify that to the next folks that come along.

andersk replied:
This is an `Enum`, not an `IntEnum`. While there are internal integer values you can get explicitly with `INVALID_REQUEST.value`, that doesn’t seem like something you’d do by accident. Meanwhile, the previous implementation could not be pickled, leading to indecipherable errors from Django’s multiprocessing-based parallel test runner when tests failed.

gnprice replied:
Well, at the very least we clearly regressed in having that invariant be expressed for anyone reading the source code, as evidenced by Alex's comment above. Originally we had:
```py
class AbstractEnum(Enum):
    """An enumeration whose members are used strictly for their names."""
    # …

class ErrorCode(AbstractEnum):
```

and that docstring was lost in 6ae5be6e75e7000dcd1855b378239491646a13c9 without getting replaced.

(FWIW writing `error_code.value` somewhere doesn't seem like that unlikely a mistake to me.)

gnprice replied:
> Meanwhile, the previous implementation could not be pickled, leading to indecipherable errors from Django’s multiprocessing-based parallel test runner when tests failed.

This is a perfect example of the sort of information that would be helpful to include in a commit message.

alexmv commented on zerver/middleware.py:
```diff
@@ -737,3 +737,26 @@ def process_request(self, request: HttpRequest) -> HttpResponse | None:
+
+class InvalidRequestError(JsonableError):
+    http_status_code: int = 403
+    code: ErrorCode = ErrorCode.INVALID_REQUEST
```
I think we should scope this error code more narrowly, and have a more specific name.  "invalid request" is _very_ broad, and we're not using it in many other cases where the request is invalid.

Joelute replied:
Updated to use BadAPIRequest and BadAPIRequestError.

alexmv commented on zerver/middleware.py:
```diff
@@ -737,3 +737,26 @@ def process_request(self, request: HttpRequest) -> HttpResponse | None:
+        return _("The Zulip API is only accessible over HTTPs.")
+
+
+class DetectInvalidRequests(MiddlewareMixin):
```
Ditto the name of this middleware.

alexmv commented on zproject/computed_settings.py:
```diff
@@ -239,6 +239,7 @@ def services_append(service_name: str) -> None:
     "django.middleware.common.CommonMiddleware",
     "zerver.middleware.LocaleMiddleware",
     "zerver.middleware.HostDomainMiddleware",
+    "zerver.middleware.DetectInvalidRequests",
     "zerver.middleware.DetectProxyMisconfiguration",
```
I believe this ordering is correct, but this is a very subtle and important part of the change.

Read the comment in DetectProxyMisconfiguration.  Its logic includes the assumption that _Django does not serve content over HTTP_, which we are breaking by making nginx pass `/api/v1` requests to Django.  I think this is _fine_, because by placing the reject-HTTP-api-requests middleware before the DetectProxyMisconfiguration, it is still the case that no HTTP requests will make it to that middleware.

Joelute replied:
Yeah I noticed that while working this so I decided it would be good to place the new middleware before `DetectProxyMisconfiguration`

alexmv reviewed (commented):
This needs some wording changes, and (most importantly) to not break existing error codes.

On the commit message -- we can make this clearer.

> Currently we are simply redirecting all HTTP requests through NGINX

I would emphasize "all unencrypted HTTP requests", since the distinction between HTTP and HTTPS may be lost at a quick read; emphasize the relevant part.

> , which is dangerous as we are sending sensitive secrets through plain/text.

"sending secrets through plain/text" is a weird phrase.  "sending secrets in cleartext" is more canonical.

"is dangerous" does not describe the danger -- we can be more explicit by saying _what_ that danger is.  Specifically, that it can expose user secrets in cleartext.

But this also does not discuss the second part of this -- that _consumers of the API may not realize this exposure due to the redirect_.  We need to include a statement that issuing a blanket redirect may cause clients to "helpfully" silently follow the redirect, leaving the users of the clients unaware that their secrets were leaked.

> We want to stop redirecting API requests and follow the fail fast principals
here to provide developers with a clear message on the issue.

("fail fast principle" not "principals", but see below)

"fast fast principals" is not really applicable here.  We're fundamentally changing the failure criteria, so it's not that the failure was too slow.  The issue was being silent when we were potentially aware that secrets had been leaked.

> We will handle unencrypted API calls with an middleware in Django with a
possibility of implementing an auto key revoker later on.

We don't need to talk about what we _might_ implement later.  Keep the message to what exists, not what might exist, unless a reader might really question a design decision without that knowledge.

The decisions here (handling errors in Django) are solid on their own merits; we don't need to provide a crystal ball.

The part of this change which is as yet unmotivated in the commit message is why we only respond to `/api/v1` requests with this, and continue to serve the redirect to all other endpoints.

> These changes employs the following:

("these changes employ" not "employs"; however, see below)

I don't like laundry lists in my commit messages.  Our [commit discipline docs](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-summary-part-2) talk about complete sentences and imperative voice, and those apply here.  It doesn't need to stick to just one sentence, but it should be one or more full sentences.

> This will allow us and developers using our API to handle the requests in a more secure manner.

"a more secure manner" doesn't say anything about what is more secure.  In general, be specific (😄) about the goal (i.e. avoiding leaking secrets), not "make it more secure."  However, we've already addressed the motivation for this commit in the first part of the commit message -- we don't need to do so again at the end.

andersk commented:
This makes no changes to the production nginx configuration, so it can’t possibly do what it says.

Joelute committed:
exceptions: Clarify usage of ErrorCode enum.

This commit reintroduce a comment that was removed in 6ae5be6 which
clarifies the usage of ErrorCode enumerator.

alexmv commented on zerver/middleware.py:
```diff
@@ -737,3 +737,31 @@ def process_request(self, request: HttpRequest) -> HttpResponse | None:
+
+class BadAPIRequest(MiddlewareMixin):
+    def process_request(self, request: HttpRequest) -> None:
+        # Most HTTP requests are redirected to HTTPS by NGINX, but this is not optimal for
+        # API requests, as it can cause security issues. This middleware is used to handle
+        # unencrypted API requests that are not made over HTTPS. Returning a easy-to-understand
+        # error message that can help clients understand the situation and fix the security
+        # issue.
```
"cause security issues" is not specific.

It is always worth re-reading the comment as if you do not know what the change does, and imagining their thought process.  This leaves me knowing it fixes ... some security problem.  But not what that problem is, or why this is the right fix.

```suggestion
        # Most HTTP requests are redirected to HTTPS by NGINX, but API
        # requests should not be, as clients might follow the redirect
        # and accidentally hide the credentials leak inherent in the
        # insecure HTTP request.  Explicitly reject HTTP requests to
        # the API.

Joelute replied:
Right, yeah I should have been more specific here as well. It just felt awkward for me to write the specifics rather than generalize.

alexmv reviewed (commented):
The commit message is better, but is still non-specific and a bit repetitive.  Here's my rewording of your commit message, which I hope you'll agree is a bit more specific and readable:

```
We currently redirect all unencrypted HTTP requests to HTTPS in NGINX.
While this is reasonable for browser-based environments, which
generally segregate HTTP and HTTPS cookies (ensuring that client
secrets are not leaked in the HTTP request), this poses an security
risk for API clients.  Specifically, an API client which is configured
to auto-follow redirects may send an API key in cleartext, and then
_silently_ follow the 301 redirect provided by NGINX, leaving the user
blind to the fact that their credentials were exposed in cleartext.

Stop redirecting unencrypted API requests, and provide clients with a
clear message alerting them to the fact that credentials were
potentially exposed.
```

Joelute committed:
dev_settings: Dynamically use URL scheme for Zulip services.

The URL scheme is currently static and set to 'http'. This change makes
use of the dynamic `EXTERNAL_URI_SCHEME` setting to use the correct
protocol.
